Userland/Libraries/LibDeviceTree/DeviceTree.cpp

```diff
@@ -93,4 +94,173 @@ bool Node::is_compatible_with(StringView wanted_compatible_string) const
     return is_compatible;
 }
 
+ErrorOr<Reg> Node::reg() const
+{
+    if (parent() == nullptr)
+        return Error::from_errno(EINVAL);
+
+    auto reg_prop = get_property("reg"sv);
+    if (!reg_prop.has_value())
+        return Error::from_errno(ENOENT);
+
+    return Reg { reg_prop->raw_data, *this };
+}
+
+ErrorOr<Ranges> Node::ranges() const
+{
+    if (parent() == nullptr)
+        return Error::from_errno(EINVAL);
+
+    auto ranges_prop = get_property("ranges"sv);
+    if (!ranges_prop.has_value())
+        return Error::from_errno(ENOENT);
+
+    return Ranges { ranges_prop->raw_data, *this };
+}
+
+ErrorOr<Address> RegEntry::resolve_root_address() const
+{
+    VERIFY(m_node.parent() != nullptr);
+    return m_node.parent()->translate_child_bus_address_to_root_address(bus_address());
+}
+
+ErrorOr<RegEntry> Reg::entry(size_t index) const
+{
+    if (index >= entry_count())
+        return Error::from_errno(EINVAL);
+
+    VERIFY(m_node.parent() != nullptr);
+
+    auto parent_address_cells = m_node.parent()->address_cells();
+    auto parent_size_cells = m_node.parent()->size_cells();
+
+    size_t const start_index = index * ((parent_address_cells + parent_size_cells) * sizeof(u32));
+
+    auto address = Address { m_raw.slice(start_index, parent_address_cells * sizeof(u32)) };
+    auto length = Size { m_raw.slice(start_index + (parent_address_cells * sizeof(u32)), parent_size_cells * sizeof(u32)) };
+
+    return RegEntry {
+        move(address),
+        move(length),
+        m_node,
+    };
+}
+
+size_t Reg::entry_count() const
+{
+    VERIFY(m_node.parent() != nullptr);
+
+    auto parent_address_cells = m_node.parent()->address_cells();
+    auto parent_size_cells = m_node.parent()->size_cells();
+
+    // #address-cells should never be 0, but still avoid dividing by 0.
+    if (parent_address_cells + parent_size_cells == 0)
+        return 0;
+
+    return m_raw.size() / ((parent_address_cells + parent_size_cells) * sizeof(u32));
+}
+
+ErrorOr<Address> RangesEntry::translate_child_bus_address_to_parent_bus_address(Address const& address) const
+{
+    auto maybe_device_type = m_node.get_property("device_type"sv);
+
+    if (maybe_device_type.has_value() && maybe_device_type->as_string() == "pci"sv) {
+        // TODO
+        return Error::from_errno(ENOTSUP);
+    }
+
+    auto address_as_flatptr = TRY(address.as_flatptr());
+    auto child_bus_address_as_flatptr = TRY(m_child_bus_address.as_flatptr());
+    auto parent_bus_address_as_flatptr = TRY(m_parent_bus_address.as_flatptr());
+    auto length_as_size_t = TRY(m_length.as_size_t());
+
+    if (address_as_flatptr >= child_bus_address_as_flatptr && address_as_flatptr < (child_bus_address_as_flatptr + length_as_size_t))
+        return Address::from_flatptr(address_as_flatptr - child_bus_address_as_flatptr + parent_bus_address_as_flatptr);
+
+    return Error::from_errno(EFAULT);
+}
```

Comment:
I know it's defacto equivalent, but I'd prefer to (also) have a check against the address-size

Reply:
Ah well, its handled by the `TRY(as_flatptr())`, so nevermind